lib/tree.rb

```diff
@@ -1,64 +1,181 @@
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    
+    return current_node
+  end
+  
+  # Time Complexity: O(log(n)) where n is the number of nodes
+  # Space Complexity: O(log(n)) where n is the number of nodes
   def find(key)
-    raise NotImplementedError
+    helper_find(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def helper_find(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+    # return nil if current_node.left.nil? && current_node.right.nil?
+    
+    if key < current_node.key
+      helper_find(current_node.left, key)
+    else
+      helper_find(current_node.right, key)
+    end
+  end
+  
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the size of the tree/number of function calls
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    # Do inorder on left subtree
+    inorder_helper(current_node.left, list)
+    
+    # Add the root
+    list << { key: current_node.key, value: current_node.value}
+    
+    # Do inorder on right subtree
+    inorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the number of nodes
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    # Add the root
+    list << { key: current_node.key, value: current_node.value}
+    
+    # Do preorder on left subtree
+    preorder_helper(current_node.left, list)
+    
+    # Do preorder on right subtree
+    preorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the number of nodes
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    # Do postorder on left subtree
+    postorder_helper(current_node.left, list)
+    
+    # Do postorder on right subtree
+    postorder_helper(current_node.right, list)
+    
+    # Add the root
+    list << { key: current_node.key, value: current_node.value}
+    
+    return list
+  end
+  
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the number of nodes
   def height
```

Comment:
You are right on time complexity since you visit each node, however your space complexity is wrong.  Because your recursive stack is never more than Log n deep.

The method isn't right as well, although it's passing the tests (and I'll add another test for next cohort). Your method is always returning `height_helper(current_node.right, max, count + 1)`, it's never checking the left side.